core/src/main/scala/kafka/controller/PartitionStateMachine.scala

```diff
@@ -125,6 +128,21 @@ class PartitionStateMachine(config: KafkaConfig,
     partitionState.filter { case (_, s) => s == state }.keySet.toSet
   }
 
+  private def changeStateTo(partition: TopicPartition, currentState: PartitionState, targetState: PartitionState): Unit = {
+    partitionState.put(partition, targetState)
+    updateControllerMetrics(partition, currentState, targetState)
+  }
+
+  private def updateControllerMetrics(partition: TopicPartition, currentState: PartitionState, targetState: PartitionState) : Unit = {
+    if (!topicDeletionManager.isTopicQueuedUpForDeletion(partition.topic)) {
+      if (currentState != OfflinePartition && targetState == OfflinePartition) {
```

Comment:
looks like we are not checking "!controllerContext.liveOrShuttingDownBrokerIds.contains()" ,  which was there is old code. Is that check not required now?

Reply:
@omkreddy Thanks for bringing this up. I discussed internally with someone in our team, and we think the check is not required now. The old definition of an offline partition is one whose leader is not alive, and the new definition is one whose state in the PartitionStateMachine is Offline. Comparing these definitions gives us 4 combinations
1. leader alive and state in not offline => both definitions count the partition as not offline
2. leader alive and state is offline
3. no alive leader and state is not offline
4. no alive leader and state is offline => both definitions count the partition as offline

In case 1 and 4 these definitions give the same result.
For case 2: by checking the code, it seems as long as there is an alive leader during events like TopicChange, BrokerFailure or ControlledShutdown, by the end of processing such events, the controller will always change the state of the partition to Online. So it seems under normal operations by the end of processing any event, case 2 will never happen. 
I also checked the failure scenario when the request to update leadership info through Zookeeper fails, then the controllerContext.partitionLeadershipInfo is not updated, and the state in PartitionStateMachine is not updated. The end result is that both definitions will use the result before the zk update.

For case 3, not being in offline state gives us 3 other possibilities, i.e. NewPartition, OnlinePartition, and NonExistentPartition. 
- NewPartition is only being used as an intermediate state, and no partition should be in this state after processing any event.
- For OnlinePartition, it should not happen when there is no alive leader.
- For NonExistentPartition, it only happens at the last step of topic deletion, at which point the leader should also be deleted. So this combination is also impossible.

In summary, the new definition should not give different results compared with the new definition by the end of processing any event. 
Please let me know if anything is not clear. Thanks!

Reply:
Thanks for the explanation. Sounds good to me.